main.go

```diff
@@ -1,30 +1,261 @@
+	Extensions extensionsCmd    `cmd:"" help:"Call all_extension_numbers_of_type"`
+}
+
+type (
+	servicesCmd struct{}
+	symbolCmd   struct {
+		Symbol string `arg:""`
+	}
+)
+
+type filenameCmd struct {
+	Filename string `arg:""`
+}
+
+type extensionCmd struct {
+	Type   string `arg:""`
+	Number int32  `arg:""`
+}
+
+type extensionsCmd struct {
+	Type string `arg:""`
 }
 
 func main() {
 	cfg := &config{}
-	_ = kong.Parse(cfg, kong.Vars{"version": version})
-	if err := run(cfg); err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		os.Exit(1)
+	kctx := kong.Parse(cfg,
+		kong.Vars{"version": version},
+		kong.Description("gRPC reflection API toolkit"),
+	)
+	err := kctx.Run(cfg.globals)
+	kctx.FatalIfErrorf(err)
+}
+
+func (cfg *config) AfterApply() error {
+	cfg.out = os.Stdout
+	if cfg.Out != "-" {
+		var err error
+		if cfg.out, err = os.Create(cfg.Out); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	cfg.hostAddress = cfg.Address
+	return nil
+}
+
+func (s *servicesCmd) Run(g globals) error {
+	stream, err := newStream(context.Background(), g)
+	if err != nil {
+		return err
+	}
+	defer stream.CloseSend()
+	req := &rpb.ServerReflectionRequest{
+		Host:           g.hostAddress,
+		MessageRequest: &rpb.ServerReflectionRequest_ListServices{},
+	}
+	resp, err := send(stream, req)
+	if err != nil {
+		return err
+	}
+	return printProto(g.out, resp, g.Format)
+}
+
+func (s *symbolCmd) Run(g globals) error {
+	stream, err := newStream(context.Background(), g)
+	if err != nil {
+		return err
+	}
+	defer stream.CloseSend()
+	req := &rpb.ServerReflectionRequest{
+		Host: g.hostAddress,
+		MessageRequest: &rpb.ServerReflectionRequest_FileContainingSymbol{
+			FileContainingSymbol: s.Symbol,
+		},
 	}
+	resp, err := send(stream, req)
+	if err != nil {
+		return err
+	}
+	return printProto(g.out, resp, g.Format)
 }
 
-func run(cfg *config) error {
+func (f *filenameCmd) Run(g globals) error {
+	stream, err := newStream(context.Background(), g)
+	if err != nil {
+		return err
+	}
+	defer stream.CloseSend()
+	req := &rpb.ServerReflectionRequest{
+		Host: g.hostAddress,
+		MessageRequest: &rpb.ServerReflectionRequest_FileByFilename{
+			FileByFilename: f.Filename,
+		},
+	}
+	resp, err := send(stream, req)
+	if err != nil {
+		return err
+	}
+	return printProto(g.out, resp, g.Format)
+}
 
-	return nil
+func (e *extensionCmd) Run(g globals) error {
+	stream, err := newStream(context.Background(), g)
+	if err != nil {
+		return err
+	}
+	defer stream.CloseSend()
+	req := &rpb.ServerReflectionRequest{
+		Host: g.hostAddress,
+		MessageRequest: &rpb.ServerReflectionRequest_FileContainingExtension{
+			FileContainingExtension: &rpb.ExtensionRequest{
+				ContainingType:  e.Type,
+				ExtensionNumber: e.Number,
+			},
+		},
+	}
+	resp, err := send(stream, req)
+	if err != nil {
+		return err
+	}
+	return printProto(g.out, resp, g.Format)
+}
+
+func (e *extensionsCmd) Run(g globals) error {
```

Comment:
All of the `Run` methods seem to be identical except for the value of `MessageRequest`. Any reason not to factor it out and just create and pass the `MessageRequest` in each `Run` method?

Reply:
yes good idea.